components/base/kinematicbase/execution.go

```
@@ -0,0 +1,471 @@
+				actualPose, err := ptgk.Localizer.CurrentPosition(ctx)
+				if err != nil {
+					return err
+				}
+				expectedPoseRel, err := ptgk.frame.Transform(currentInputs)
+				if err != nil {
+					return err
+				}
+
+				// This is where we expected to be on the trajectory
+				expectedPose := spatialmath.Compose(step.arcSegment.StartPosition, expectedPoseRel)
+
+				// This is where actually are on the trajectory
+				poseDiff := spatialmath.PoseBetween(actualPose.Pose(), expectedPose)
+
+				allowableDiff := ptgk.linVelocityMMPerSecond * inputUpdateStepSeconds * (minDeviationToCorrectPct / 100)
+				ptgk.logger.Debug("allowable diff ", allowableDiff, " diff now ", poseDiff.Point().Norm())
+				if poseDiff.Point().Norm() > allowableDiff || poseDiff.Orientation().AxisAngles().Theta > 0.25 {
+					ptgk.logger.Debug("expected to be at ", spatialmath.PoseToProtobuf(expectedPose))
+					ptgk.logger.Debug("Localizer says at ", spatialmath.PoseToProtobuf(actualPose.Pose()))
+					// Accumulate list of points along the path to try to connect to
+					goals := ptgk.makeCourseCorrectionGoals(
+						goalsToAttempt,
+						actualPose.Pose(),
+					)
+
+					ptgk.logger.Debug("wanted to attempt ", goalsToAttempt, " goals, got ", len(goals))
+					// Attempt to solve from `actualPose` to each of those points
+					solution, err := ptgk.courseCorrect(ctx, goals)
+					if err != nil {
+						ptgk.logger.Debug(err)
+					}
+					if solution.Solution != nil {
+						ptgk.logger.Debug("successful course correction", solution.Solution)
+
+						correctiveArcSteps := []arcStep{}
+						for i := 0; i < len(solution.Solution); i += 2 {
+							// We've got a course correction solution. Swap out the relevant arcsteps.
+							correctiveTraj, err := ptgk.ptgs[ptgk.courseCorrectionIdx].Trajectory(
+								solution.Solution[i].Value,
+								solution.Solution[i+1].Value,
+								stepDistResolution,
+							)
+							if err != nil {
+								ptgk.logger.Warn(err)
```

[q]: why is ok to ignore error here, wouldn't we be dealing with non-sense `trajNodes`?